kvserver: de-flake TestReplicateReAddAfterDown

tbg opened the pull request:
When restarting a server, it's on a different port, which can give
spurious connection errors for a moment (but which the system itself
cannot recognize as spurious). Teach the test harness to avoid those,
to eliminate this general category of test flake.

Closes #59453.

Release note: None

tbg committed:
kvserver: de-flake TestReplicateReAddAfterDown

When restarting a server, it's on a different port, which can give
spurious connection errors for a moment (but which the system itself
cannot recognize as spurious). Teach the test harness to avoid those,
to eliminate this general category of test flake.

Closes #59453.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59795)
<!-- Reviewable:end -->

knz reviewed (approved):
that is a clever fix. Thanks!

lunevalex reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59795#-:-MShbjUT5J1Ry_cctsKP:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=lunevalex,knz

tbg commented:
bors r=lunevalex,knz

tbg commented:
My debugger tells me the error comes from here:

https://github.com/cockroachdb/vendored/blob/e1ac11725750ec09d1b76a76a8f77c0fa7b8f71c/google.golang.org/grpc/picker_wrapper.go#L141

tbg commented:
Oh boy. I think this to some degree user (my) error, but also, huge footgun?

When I added the code to testcluster, I added a `cc.Close()` - this closes the connection but it's still in the rpc context's pool, so... things are bad?

Amazingly, it doesn't look like we do this wrong anywhere in prod code - I would've expected to see errant `defer cc.Close()`s everywhere but I don't.

tbg commented:
bors r=lunevalex,knz
